### PR TITLE
Disconnected support

### PR DESCRIPTION
**Pull request by jaredoconnell**
### Description
This PR successfully makes it so the dashboard templates are loaded into the syncer image, so that they can be deployed at runtime without cloning this repository.

The downside to this change is the syncer image needs to be updated for each PR, so that must be setup on quay.io

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -20,7 +20,9 @@ $(ALLDIRS):
 format: deps
 	$(BINDIR)/jsonnetfmt -i $(TEMPLATES)
 
-build: deps $(TEMPLATESDIR)/grafonnet-lib $(outputs)
+build: deps $(TEMPLATESDIR)/grafonnet-lib offline-build
```
Comment (rsevilla87) [inline Makefile:23]: Why did you add offline-build to this makefile target?

File: dittybopper/syncer/Dockerfile [hunk 2]
```diff
@@ -2,6 +2,8 @@ FROM registry.access.redhat.com/ubi8/ubi-minimal
 
 RUN microdnf install git make tar
 WORKDIR /dittybopper-syncer
-RUN chmod 775 /dittybopper-syncer
-COPY entrypoint.sh /bin/entrypoint.sh
+COPY . /dittybopper-syncer/dashboards
+COPY dittybopper/syncer/entrypoint.sh /bin/entrypoint.sh
```
Comment (rsevilla87) [inline dittybopper/syncer/Dockerfile:6]: This container build is not working for me...
```
STEP 1/8: FROM registry.access.redhat.com/ubi8/ubi-minimal
STEP 2/8: RUN microdnf install git make tar
--> Using cache d3d21a05b29fa1617f0b1c6ca0556bdd33db2d50f013f9d7b67524c102f0a342
--> d3d21a05b29
STEP 3/8: WORKDIR /dittybopper-syncer
--> Using cache bdee90ad2fa013d1f11dd728659eff56142ee95921363fa6e3ee394f63ffbccf
--> bdee90ad2fa
STEP 4/8: COPY . /dittybopper-syncer/dashboards
--> 1e9205394b7
STEP 5/8: COPY dittybopper/syncer/entrypoint.sh /bin/entrypoint.sh
Error: error building at STEP "COPY dittybopper/syncer/entrypoint.sh /bin/entrypoint.sh": checking on sources under "******/performance-dashboards/dittybopper/syncer": copier: stat: "/dittybopper/syncer/entrypoint.sh": no such file or directory

``

**Review by rsevilla87 — CHANGES_REQUESTED** [review]
This PR is a bit confusing IMO.
With the new changes, the syncer image expects the dashboard to be rendered in the localhost before copying them into the image, I think we should take advantage of a multi-stage Dockerfiles to build this container image and prevent pulling dependencies in runtime and building the dashboards in localhost:
i.e


```Dockerfile
FROM registry.access.redhat.com/ubi8/ubi-minimal as builder
RUN microdnf install git make tar
RUN git clone https://github.com/cloud-bulldozer/performance-dashboards.git --depth=1
RUN make -C performance-dashboards build

FROM registry.access.redhat.com/ubi8/ubi-minimal
WORKDIR /dittybopper-syncer
RUN chmod 775 /dittybopper-syncer
COPY --from=builder /performance-dashboards/rendered/* .
COPY entrypoint.sh /bin/entrypoint.sh
ENTRYPOINT ["entrypoint.sh"]
```

File: dittybopper/syncer/Dockerfile (hunk 2 above)
Comment (jaredoconnell) [reply dittybopper/syncer/Dockerfile:6]: I moved the build path to the root of the repository. Is that where you built it from?

**Comment by jaredoconnell** [comment]
> This PR is a bit confusing IMO. With the new changes, the syncer image expects the dashboard to be rendered in the localhost before copying them into the image, I think we should take advantage of a multi-stage Dockerfiles to build this container image and prevent pulling dependencies in runtime and building the dashboards in localhost: i.e
> 
> ```dockerfile
> FROM registry.access.redhat.com/ubi8/ubi-minimal as builder
> RUN microdnf install git make tar
> RUN git clone https://github.com/cloud-bulldozer/performance-dashboards.git --depth=1
> RUN make -C performance-dashboards build
> 
> FROM registry.access.redhat.com/ubi8/ubi-minimal
> WORKDIR /dittybopper-syncer
> RUN chmod 775 /dittybopper-syncer
> COPY --from=builder /performance-dashboards/rendered/* .
> COPY entrypoint.sh /bin/entrypoint.sh
> ENTRYPOINT ["entrypoint.sh"]
> ```

I will change it to a multi-stage build to remove the Git dependency. Do you want the jsonnet rendered on container build, increasing image size, or when it's run, reducing it? Unless the make dependency exceeds the size of the rendered jsonnet.

**Comment by rsevilla87** [comment]
> > This PR is a bit confusing IMO. With the new changes, the syncer image expects the dashboard to be rendered in the localhost before copying them into the image, I think we should take advantage of a multi-stage Dockerfiles to build this container image and prevent pulling dependencies in runtime and building the dashboards in localhost: i.e
> > ```dockerfile
> > FROM registry.access.redhat.com/ubi8/ubi-minimal as builder
> > RUN microdnf install git make tar
> > RUN git clone https://github.com/cloud-bulldozer/performance-dashboards.git --depth=1
> > RUN make -C performance-dashboards build
> > 
> > FROM registry.access.redhat.com/ubi8/ubi-minimal
> > WORKDIR /dittybopper-syncer
> > RUN chmod 775 /dittybopper-syncer
> > COPY --from=builder /performance-dashboards/rendered/* .
> > COPY entrypoint.sh /bin/entrypoint.sh
> > ENTRYPOINT ["entrypoint.sh"]
> > ```
> 
> I will change it to a multi-stage build to remove the Git dependency. Do you want the jsonnet rendered on container build, increasing image size, or when it's run, reducing it? Unless the make dependency exceeds the size of the rendered jsonnet.

Only the "build" container size will increase, the resulting container image will be small

**Comment by jaredoconnell** [comment]
> > > This PR is a bit confusing IMO. With the new changes, the syncer image expects the dashboard to be rendered in the localhost before copying them into the image, I think we should take advantage of a multi-stage Dockerfiles to build this container image and prevent pulling dependencies in runtime and building the dashboards in localhost: i.e
> > > ```dockerfile
> > > FROM registry.access.redhat.com/ubi8/ubi-minimal as builder
> > > RUN microdnf install git make tar
> > > RUN git clone https://github.com/cloud-bulldozer/performance-dashboards.git --depth=1
> > > RUN make -C performance-dashboards build
> > > 
> > > FROM registry.access.redhat.com/ubi8/ubi-minimal
> > > WORKDIR /dittybopper-syncer
> > > RUN chmod 775 /dittybopper-syncer
> > > COPY --from=builder /performance-dashboards/rendered/* .
> > > COPY entrypoint.sh /bin/entrypoint.sh
> > > ENTRYPOINT ["entrypoint.sh"]
> > > ```
> > 
> > 
> > I will change it to a multi-stage build to remove the Git dependency. Do you want the jsonnet rendered on container build, increasing image size, or when it's run, reducing it? Unless the make dependency exceeds the size of the rendered jsonnet.
> 
> Only the "build" container size will increase, the resulting container image will be small

It will also increase by the size of the rendered dashboards, but I'm guessing that's not too much in the greater scheme of things.

**Comment by jaredoconnell** [comment]
The latest commit splits it into a two stage build. The new container is now less than half the size of the old one, and it no longer has any security vulnerabilities.
It needs to be built from the root of the project.

File: dittybopper/syncer/Dockerfile [hunk 3]
```diff
@@ -1,7 +1,12 @@
-FROM registry.access.redhat.com/ubi8/ubi-minimal
+FROM registry.access.redhat.com/ubi8/ubi-minimal AS builder
+RUN microdnf install make git tar
+WORKDIR /performance-dashboards
+COPY . /performance-dashboards
+RUN make deps build
```
Comment (rsevilla87) [inline dittybopper/syncer/Dockerfile:5]: This can be replaced by just `make`

File: dittybopper/README.md [hunk 4]
```diff
@@ -19,6 +19,18 @@ templates will need adjustment accordingly.
...
+The image is built with the context at the root of the repository, and the image in the dittybopper/syncer directory.
+
+For disconnected support, the syncer image stores all dashboards on it. For deploying dittybopper with changed
+dashboards, you need to build it yourself from the root of the repository and update the SYNCER_IMAGE environment
+variable to match your own image repository.
+
+If using disconnected, you need to sync the cloud-bulldozer grafana image (shown in the
+dittybopper/templates/dittybopper.yaml.template file) and your chosen syncer image
+(defaults to quay.io/cloud-bulldozer/dittybopper-syncer:latest).
+
```
Comment (rsevilla87) [inline dittybopper/README.md:33]: Im missing some instructions about how to build your own syncer image, i.e:

```suggestion
The syncer image can be built by executing the following command from folder of this repository.

podman build -f dittybopper/syncer/Dockerfile -t=container.registry.org/organization/syncer:latest  .
```

**Review by rsevilla87 — APPROVED** [review]
LGTM!, good stuff @jaredoconnell
